### PR TITLE
Add replacement implementation of Lua IO

### PR DESCRIPTION
**Pull request by wingo**
The goal is to replace Lua's "io" library with a compatible interface, except written in Lua instead of C, and allowing extra functionality such as string files, mmap buffers, and non-blocking concurrency.

Eventually this will allow for a number of refactorings:

 * `lib.ctable` will be able to use these standard streams for its serialization and deserialization
 * `lib.yang.binary` likewise
 * `lib.yang.parser` will be able to take a stream as its argument instead of a string
 * `lib.yang.data` too, we get to remove `lib.yang.stream`
 * `lib.ptree.json` likewise
 * we'll be able to use fibers when needed, perhaps in the `lib.ptree` manager, simplifying parsing and serialization of RPCs; the stream abstraction is ready for cooperative coroutine-based multitasking and integration with an epoll loop
 * `snabb config` will be able to stream input and output, avoiding the need to build enormous strings for large configurations (i.e. large routing tables)

I think in the end we will have more functionality with less code and still good speed.

The current status is that the `file` methods are complete but the `io` module is not yet done.

### CODE REVIEW

File: src/lib/buffer.lua [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+   if count > self:read_avail() then error('read xrun') end
+   local pos = self:read_pos()
+   local count1 = math.min(self.size - pos, count)
+   ffi.copy(bytes, self.buf + pos, count1)
+   ffi.copy(bytes + count1, self.buf, count - count1)
+   self:advance_read(count)
+end
+
+function buffer:drop()
+   if count > self:read_avail() then error('read xrun') end
```
Comment (dpino) [inline src/lib/buffer.lua]: `count` is not defined.

**Review by dpino — COMMENTED** [review]
Looking good!

**Comment by wingo** [comment]
I rebased this commit to be based on `next` instead of `lwaftr`, and fed in some feedback from fibers, which I also rebased on top of this work.  Apologies for any confusion!

**Comment by wingo** [comment]
After a couple more rebases, this is ready to go.

Everything is streaming except the RPC protocol, which has a couple of cases in which it serializes configurations as strings.  One is in the YANG side:

```yang
  rpc get-schema {
    input {
      leaf schema { type string; mandatory true; }
      leaf revision { type string; }
    }
    output {
      uses error-reporting;
      leaf source { type string; }
    }
  }
```

I.e. the `source` is a string.  Probably the fix here is to use `anydata` from YANG 1.1.

The other aspect that's a problem is the RPC protocol itself, which consists of messages with a length prefix.  We could instead use a strategy like HTTP's "chunked" transfer encoding.  Future work in any case.

**Review by dpino — APPROVED** [review]

**Pull request merged**